### PR TITLE
[Inductor] record time for 'compile time' autotuning

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139431

Here are the cases that Inductor does autotuning at compile time:
1. pad mm: benchmark to decide if we should pad or not
2. template autotuning: benchmark triton/cutlass templates and ATen  kernel for matmul/conv and pick the fastest one.

The PR annotate these cases with `dynamo_timed`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Comment by ezyang** [comment]
It is possible this is right but I have some questions about the overall code flow in this area. It would be also useful to run with this logic on a representative model that triggers these autotuning and let is see the perfetto where these would show up.

Questions:
* I see some comments about subprocess tuning; is the instrumentation here always called in process or can it be out of process
* Is the autotune function called only once or multiple times?

A slight improvement would be to record some basic metadata about what tuning has taken place

File: torch/_inductor/select_algorithm.py [hunk 1]
```diff
@@ -1338,7 +1338,8 @@ def wait_on_futures():
             return wait_on_futures
 
         def autotune(choices):
-            return make_benchmark_fn()(choices)
+            with dynamo_timed("template_autotuning"):
+                return make_benchmark_fn()(choices)
```
Comment (eellison) [inline torch/_inductor/select_algorithm.py:1342]: Have you spot checked this ? does it cover the benchmarking and the compilation ? i'm not exactly sure how we should record the precompilation, since it is async..

Comment (eellison) [reply torch/_inductor/select_algorithm.py:1342]: I think we also want the time in `benchmark_fused_nodes`

**Review by eellison — COMMENTED** [review]
couple comments

File: torch/_inductor/select_algorithm.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/select_algorithm.py:1342]: I created a toy script to do matmul and spot check the result as:
```
(Pdb) [U.compilation_time_metrics[key] for key in ["pad_mm_benchmark", "template_autotuning"]]
[[4.7206878662109375e-05], [2.0357143878936768]]
```

This should only cover the autotuning. Precompilation is parallel but we do wait for all `futures` to be fulfilled at some point. We can probably log that wait time

**Comment by shunting314** [comment]
> I see some comments about subprocess tuning; is the instrumentation here always called in process or can it be out of process

This instrumentation always happens in the current process. We decide if we should do the benchmarking in a child process in `make_benchmark_fn` which is called by the `autotune` method.

> Is the autotune function called only once or multiple times?

It's called only once for a single matmul invocation. But there can be one such call for each matmul, conv calls: e.g., if a model has 2 matmul, one addmm, 3 conv: we should see this function being called 6 times.

> A slight improvement would be to record some basic metadata about what tuning has taken place

I can add the benchmark name (i.e. matmul/addmm/conv) to the metric key. Not sure if there is a better way to log extra metadata.

> It would be also useful to run with this logic on a representative model that triggers these autotuning and let is see the perfetto where these would show up.

I guess I need some pointer about how to do this, lol

File: torch/_inductor/select_algorithm.py (hunk 1 above)
Comment (ezyang) [reply torch/_inductor/select_algorithm.py]: I have no calibration for whether or not these figures are correct or not. Do these match your expectations? In particular pad_mm_benchmark seems really fast!

**Comment by ezyang** [comment]
> I guess I need some pointer about how to do this, lol

You have a model right? Just run it with your changes and TORCH_TRACE=/tmp/a and then tlparse it

**Comment by ezyang** [comment]
As for if these are the right spots, one thing I can offer is to code read all of these sites and convince myself they are right. Would you like me to do that or is there a faster approach?

**Comment by shunting314** [comment]
> You have a model right? Just run it with your changes and TORCH_TRACE=/tmp/a and then tlparse it

I thought I need leverage some meta internal infra. Good to know that tlparse can make it.

File: torch/_inductor/select_algorithm.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/select_algorithm.py]: pad_mm_benchmark will be fast if there are no unaligned mms to benchmark.

Comment (shunting314) [reply torch/_inductor/select_algorithm.py]: I think it's expected since the tensor sizes I pick are quite regular. Some heuristics probably skips benchmarking for padding and make it super fast

Comment (shunting314) [reply torch/_inductor/select_algorithm.py]: I think it's due to this check in mm padding: https://github.com/pytorch/pytorch/blob/ff616c26fb66b26966b6f6f50aa85bf615400816/torch/_inductor/fx_passes/pad_mm.py#L413 .  cc @ezyang

**Comment by shunting314** [comment]
Here is chromium trace: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpYOfugN/chromium_events.json when training BertForMaskedLM with MaxAutotune enabled.

Spot checking the timeline, it seems to be reasonable.
Screenshot 2024-11-06 at 1 55 49 PM

- The leftmost green box is for mm-padding
- we also see benchmarking for matmul and fused nodes in `scheduler.__init__`

**Review by ezyang — APPROVED** [review]
Holy smokes, this is a lot! Thanks!

@masnesral there is a lot of volume going on new dynamo_timed sites, you need to make sure you don't change the default or you will hit tons of silent land races
